homeassistant/loader.py

```diff
@@ -42,133 +40,92 @@
-    # Because of this behavior, we will approach loading sub components
-    # with caution: only load it if we can verify that the parent exists.
-    # We do not want to silent the ImportErrors as they provide valuable
-    # information to track down when debugging Home Assistant.
-
-    # First check custom, then built-in
-    potential_paths = ['custom_components.{}'.format(comp_name),
-                       'homeassistant.components.{}'.format(comp_name)]
-
-    for path in potential_paths:
-        # Validate here that root component exists
-        # If path contains a '.' we are specifying a sub-component
-        # Using rsplit we get the parent component from sub-component
-        root_comp = path.rsplit(".", 1)[0] if '.' in comp_name else path
-
-        if root_comp not in AVAILABLE_COMPONENTS:
-            continue
+    # Try custom component
+    module = _load_module(hass.config.path(PATH_CUSTOM_COMPONENTS),
+                          comp_or_platform)
 
+    if module is None:
         try:
-            module = importlib.import_module(path)
+            module = importlib.import_module(
+                '{}.{}'.format(PACKAGE_COMPONENTS, comp_or_platform))
+        except ImportError:
+            module = None
 
-            # In Python 3 you can import files from directories that do not
-            # contain the file __init__.py. A directory is a valid module if
-            # it contains a file with the .py extension. In this case Python
-            # will succeed in importing the directory as a module and call it
-            # a namespace. We do not care about namespaces.
-            # This prevents that when only
-            # custom_components/switch/some_platform.py exists,
-            # the import custom_components.switch would succeed.
-            if module.__spec__.origin == 'namespace':
-                continue
+    cache = hass.data.get(DATA_KEY)
+    if cache is None:
+        cache = hass.data[DATA_KEY] = {}
+    cache[comp_or_platform] = module
 
-            _LOGGER.info("Loaded %s from %s", comp_name, path)
+    return module
 
-            _COMPONENT_CACHE[comp_name] = module
 
-            return module
+def _find_spec(path, name):
+    for finder in sys.meta_path:
+        try:
+            spec = finder.find_spec(name, path=path)
+            if spec is not None:
+                return spec
+        except AttributeError:
+            # Not all finders have the find_spec method
+            pass
+    return None
 
-        except ImportError as err:
-            # This error happens if for example custom_components/switch
-            # exists and we try to load switch.demo.
-            if str(err) != "No module named '{}'".format(path):
-                _LOGGER.exception(
-                    ("Error loading %s. Make sure all "
-                     "dependencies are installed"), path)
 
-    _LOGGER.error("Unable to find component %s", comp_name)
+def _load_module(path, name):
```

Comment:
This didn't make much sense for me initially either, this blog post really helped: https://manikos.github.io/how-pythons-import-machinery-works